File: internal/gossip/bootstrap.go
package gossip

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"time"

	"repram/internal/logging"
)

// BootstrapRequest is sent when a node wants to join the cluster
type BootstrapRequest struct {
	NodeID      string `json:"node_id"`
	Address     string `json:"address"`
	GossipPort  int    `json:"gossip_port"`
	HTTPPort    int    `json:"http_port"`
}

// BootstrapResponse contains the current cluster topology
type BootstrapResponse struct {
	Success bool    `json:"success"`
	Peers   []*Node `json:"peers"`
}

// Bootstrap connects to seed nodes and retrieves the cluster topology
func (p *Protocol) Bootstrap(ctx context.Context, seedNodes []string) error {
	logging.Info("[%s] Starting bootstrap process with %d seed nodes", p.localNode.ID, len(seedNodes))

	req := &BootstrapRequest{
		NodeID:     string(p.localNode.ID),
		Address:    p.localNode.Address,
		GossipPort: p.localNode.Port,
		HTTPPort:   p.localNode.HTTPPort,
	}

	// Try each seed node until we get a successful response
	for _, seed := range seedNodes {
		logging.Debug("[%s] Attempting to bootstrap from %s", p.localNode.ID, seed)

		peers, err := p.sendBootstrapRequest(ctx, seed, req)
		if err != nil {
			logging.Warn("[%s] Failed to bootstrap from %s: %v", p.localNode.ID, seed, err)
			continue
		}

		// Add all discovered peers
		for _, peer := range peers {
			if peer.ID != p.localNode.ID {
				p.addPeer(peer)
				logging.Info("[%s] Discovered peer %s via bootstrap", p.localNode.ID, peer.ID)
			}
		}

		logging.Info("[%s] Bootstrap successful, discovered %d peers", p.localNode.ID, len(peers))
		return nil
	}

	// If no seed nodes responded, we might be the first node
	logging.Info("[%s] No seed nodes available, starting as first node", p.localNode.ID)
	return nil
}

func (p *Protocol) sendBootstrapRequest(ctx context.Context, seedAddr string, req *BootstrapRequest) ([]*Node, error) {
	jsonData, err := json.Marshal(req)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal request: %w", err)
	}

	url := fmt.Sprintf("http://%s/v1/bootstrap", seedAddr)
	httpReq, err := http.NewRequestWithContext(ctx, "POST", url, bytes.NewBuffer(jsonData))
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}
	httpReq.Header.Set("Content-Type", "application/json")
	if p.clusterSecret != "" {
		httpReq.Header.Set("X-Repram-Signature", SignBody(p.clusterSecret, jsonData))
	}

	client := &http.Client{Timeout: 5 * time.Second}
	resp, err := client.Do(httpReq)
	if err != nil {
		return nil, fmt.Errorf("failed to send request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("bootstrap rejected with status: %d", resp.StatusCode)
	}

	var bootstrapResp BootstrapResponse
	if err := json.NewDecoder(resp.Body).Decode(&bootstrapResp); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	if !bootstrapResp.Success {
		return nil, fmt.Errorf("bootstrap failed")
	}

	return bootstrapResp.Peers, nil
}

// HandleBootstrap processes incoming bootstrap requests
func (p *Protocol) HandleBootstrap(req *BootstrapRequest) *BootstrapResponse {
	// Create node info from request
	newNode := &Node{
		ID:       NodeID(req.NodeID),
		Address:  req.Address,
		Port:     req.GossipPort,
		HTTPPort: req.HTTPPort,
	}

	// Add the new node as a peer
	p.addPeer(newNode)
	logging.Info("[%s] Node %s joined via bootstrap", p.localNode.ID, req.NodeID)

	// Notify all existing peers about the new node
	// This ensures all nodes know about each other
	go p.notifyPeersAboutNewNode(newNode)

	// Return current cluster topology
	peers := p.getPeers()
	// Include ourselves in the response
	allPeers := append(peers, p.localNode)

	return &BootstrapResponse{
		Success: true,
		Peers:   allPeers,
	}
}

// notifyPeersAboutNewNode sends a SYNC message to existing peers about a new node
func (p *Protocol) notifyPeersAboutNewNode(newNode *Node) {
	peers := p.getPeers()
	for _, peer := range peers {
		if peer.ID != newNode.ID {
			// Create a SYNC message with the new node info
			msg := &Message{
				Type:      MessageTypeSync,
				From:      p.localNode.ID,
				Timestamp: time.Now(),
				MessageID: generateMessageID(),
				NodeInfo:  newNode,
			}

			// Retry SYNC messages with exponential backoff
			go p.sendSyncWithRetry(peer, msg, newNode.ID, 3)
		}
	}
}

// sendSyncWithRetry attempts to send a SYNC message with exponential backoff retry
func (p *Protocol) sendSyncWithRetry(peer *Node, msg *Message, newNodeID NodeID, maxRetries int) {
	for attempt := 0; attempt < maxRetries; attempt++ {
		if err := p.Send(context.Background(), peer, msg); err != nil {
			if attempt == maxRetries-1 {
				logging.Error("[%s] Failed to notify %s about new node %s after %d attempts: %v",
					p.localNode.ID, peer.ID, newNodeID, maxRetries, err)
			} else {
				// Exponential backoff: 1s, 2s, 4s
				delay := time.Duration(1<<attempt) * time.Second
				logging.Warn("[%s] Failed to notify %s about new node %s (attempt %d/%d), retrying in %v: %v",
					p.localNode.ID, peer.ID, newNodeID, attempt+1, maxRetries, delay, err)
				time.Sleep(delay)
			}
		} else {
			logging.Debug("[%s] Notified %s about new node %s (attempt %d)",
				p.localNode.ID, peer.ID, newNodeID, attempt+1)
			return
		}
	}
}

File: repram-mcp/src/node/bootstrap.ts
/**
 * Bootstrap — signed-root-list resolution and cluster join handshake.
 *
 * Port of internal/gossip/bootstrap.go and cmd/repram/main.go's
 * resolveOmegaBootstrap. The unsigned SRV/A-based bootstrap was removed
 * as part of REPRAM 2.1 (see docs/REPRAM-2.1-Spec.md).
 */

import { signBody } from "./auth.js";
import type { Logger } from "./logger.js";
import type { NodeInfo } from "./types.js";
import { fetchSigned } from "./trust/resolver.js";
import { loadCache, saveCache, defaultCacheDir } from "./trust/cache.js";
import { publicKeyFromBase64, SignedList } from "./trust/signed-list.js";
import { OMEGA_PUBKEY } from "./trust/omega.js";

// --- Wire format for bootstrap handshake ---

export interface BootstrapRequest {
  node_id: string;
  address: string;
  gossip_port: number;
  http_port: number;
  enclave?: string;
}

export interface BootstrapResponse {
  success: boolean;
  peers: WireBootstrapPeer[];
}

interface WireBootstrapPeer {
  id: string;
  address: string;
  port: number;       // gossip port
  http_port: number;
  enclave?: string;
}

// --- Omega signed-root-list discovery ---

/**
 * resolveOmegaBootstrap returns a verified signed root list for the public
 * network, fetching over DNS with on-disk cache fallback. Rejects when no
 * valid trust anchor is available — a node that cannot verify must not
 * join the public network.
 *
 * Callers use `result.nodes` as the seed peer list and additionally check
 * `result` against their own advertised address to determine root status.
 */
export async function resolveOmegaBootstrap(logger: Logger): Promise<SignedList> {
  const pubKey = publicKeyFromBase64(OMEGA_PUBKEY);
  const cacheDir = defaultCacheDir();
  const now = new Date();

  try {
    const list = await fetchSigned({}, pubKey, now);
    try {
      await saveCache(cacheDir, list);
    } catch (err) {
      logger.warn(`Failed to update omega cache at ${cacheDir}: ${err} (continuing)`);
    }
    logger.info(
      `Omega bootstrap: verified signed root list (${list.nodes.length} nodes, expires ${new Date(list.expires * 1000).toISOString()})`,
    );
    return list;
  } catch (fetchErr) {
    logger.warn(`Omega DNS fetch failed: ${fetchErr} — trying cached list at ${cacheDir}`);
    const cached = await loadCache(cacheDir);
    if (!cached) {
      throw new Error(`no DNS record and no cache available: ${(fetchErr as Error).message}`);
    }
    const verifyErr = cached.verify(pubKey, now);
    if (verifyErr) {
      throw new Error(`cached omega list invalid: ${verifyErr.message} (dns: ${(fetchErr as Error).message})`);
    }
    logger.info(
      `Omega bootstrap: using cached signed root list (${cached.nodes.length} nodes, expires ${new Date(cached.expires * 1000).toISOString()})`,
    );
    return cached;
  }
}

// --- Bootstrap handshake ---

/**
 * Contacts ALL seed nodes (not just the first responder) so every seed
 * registers the joining node directly. Stopping at the first success
 * leaves later seeds unaware of the joiner until topology sync, producing
 * asymmetric topologies (#82, F3). Peers from each response are
 * deduplicated by node ID across all seeds; self is always filtered out
 * (#82, F4).
 */
export async function bootstrapFromPeers(
  seedPeers: string[],
  localNode: NodeInfo,
  clusterSecret: string,
  logger: Logger,
): Promise<NodeInfo[]> {
  logger.info(`Starting bootstrap process with ${seedPeers.length} seed nodes`);

  const request: BootstrapRequest = {
    node_id: localNode.id,
    address: localNode.address,
    gossip_port: localNode.port,
    http_port: localNode.httpPort,
  };
  if (localNode.enclave) {
    request.enclave = localNode.enclave;
  }

  const seen = new Map<string, NodeInfo>();
  let successfulSeeds = 0;

  for (const seed of seedPeers) {
    try {
      const peers = await sendBootstrapRequest(seed, request, clusterSecret, logger);
      successfulSeeds++;
      for (const peer of peers) {
        if (peer.id === localNode.id) continue;
        if (seen.has(peer.id)) continue;
        seen.set(peer.id, peer);
      }
    } catch (err) {
      logger.warn(`Failed to bootstrap from ${seed}: ${err}`);
    }
  }

  if (successfulSeeds === 0) {
    logger.info("No seed nodes available, starting as first node");
    return [];
  }

  const discovered = Array.from(seen.values());
  logger.info(
    `Bootstrap complete: contacted ${successfulSeeds}/${seedPeers.length} seeds, ` +
      `discovered ${discovered.length} unique peers`,
  );
  return discovered;
}

async function sendBootstrapRequest(
  seedAddr: string,
  request: BootstrapRequest,
  clusterSecret: string,
  logger: Logger,
): Promise<NodeInfo[]> {
  const jsonBody = JSON.stringify(request);
  const bodyBuffer = Buffer.from(jsonBody);

  const url = `http://${seedAddr}/v1/bootstrap`;
  const headers: Record<string, string> = {
    "Content-Type": "application/json",
  };

  if (clusterSecret) {
    headers["X-Repram-Signature"] = signBody(clusterSecret, bodyBuffer);
  }

  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), 5_000);

  try {
    const response = await fetch(url, {
      method: "POST",
      headers,
      body: jsonBody,
      signal: controller.signal,
    });

    if (!response.ok) {
      throw new Error(`bootstrap rejected with status ${response.status}`);
    }

    const body = (await response.json()) as BootstrapResponse;
    if (!body.success) {
      throw new Error("bootstrap failed");
    }

    return (body.peers ?? []).map(wirePeerToNodeInfo);
  } finally {
    clearTimeout(timeout);
  }
}

function wirePeerToNodeInfo(peer: WireBootstrapPeer): NodeInfo {
  return {
    id: peer.id,
    address: peer.address,
    port: peer.port,
    httpPort: peer.http_port,
    enclave: peer.enclave ?? "default",
  };
}

// --- Notify peers about new node (fire-and-forget with retry) ---

export async function notifyPeerAboutNewNode(
  peerAddr: string,
  request: BootstrapRequest,
  clusterSecret: string,
  logger: Logger,
  maxRetries: number = 3,
): Promise<void> {
  const jsonBody = JSON.stringify(request);
  const bodyBuffer = Buffer.from(jsonBody);

  for (let attempt = 0; attempt < maxRetries; attempt++) {
    try {
      const url = `http://${peerAddr}/v1/bootstrap`;
      const headers: Record<string, string> = {
        "Content-Type": "application/json",
      };
      if (clusterSecret) {
        headers["X-Repram-Signature"] = signBody(clusterSecret, bodyBuffer);
      }

      const controller = new AbortController();
      const timeout = setTimeout(() => controller.abort(), 5_000);

      try {
        const response = await fetch(url, {
          method: "POST",
          headers,
          body: jsonBody,
          signal: controller.signal,
        });

        if (response.ok) {
          logger.debug(`Notified ${peerAddr} about new node (attempt ${attempt + 1})`);
          return;
        }
      } finally {
        clearTimeout(timeout);
      }
    } catch (err) {
      if (attempt === maxRetries - 1) {
        logger.error(`Failed to notify ${peerAddr} after ${maxRetries} attempts: ${err}`);
      } else {
        const delay = (1 << attempt) * 1000; // 1s, 2s, 4s
        logger.warn(`Failed to notify ${peerAddr} (attempt ${attempt + 1}/${maxRetries}), retrying in ${delay}ms`);
        await sleep(delay);
      }
    }
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}
